lib/rules/no-extra-boolean-cast.js

```diff
@@ -91,7 +91,14 @@ module.exports = {
                     context.report({
                         node,
                         message: "Redundant Boolean call.",
-                        fix: fixer => fixer.replaceText(node, sourceCode.getText(node.arguments[0]))
+                        fix: fixer => {
+                            const argument = node.arguments[0];
+
+                            if (argument.type === "LogicalExpression") {
```

Comment:
This will fix the specific case in the bug report, but I think the problem will still exist for arguments that aren't `LogicalExpression` nodes. For example, this fix would still be incorrect:

```js
!Boolean(1 + 2);

// gets fixed to:

!1 + 2;
```

I think best solution would be to check the *precedence* of the argument. The fixer should only add parentheses if the argument has lower precedence than the `!Boolean(...)` node, which is a `UnaryExpression`. (If the argument has higher precedence, the `!` operator will get applied to the entire argument anyway.)

For example, a `UnaryExpression` has a higher precedence than a `LogicalExpression`, so `!foo && bar` gets parsed as `(!foo) && bar`. On the other hand, a `CallExpression` has lower precedence than a `UnaryExpression`, so `!foo()` will get parsed as `!(foo())` rather than `(!foo)()`.

A function to get the precedence of a node can be found [here](https://github.com/eslint/eslint/blob/427543afb4bf024869b8a243c793824eca73ad67/lib/ast-utils.js#L613-L703) in `ast-utils.js`